### ISSUE TITLE
Enphase Microinverter mapping

### ISSUE THREAD

**Issue by SumDum** [domoticz/domoticz#5555]
Domoticz 14975 and 14980 running on rPi Zero W 32bit Raspbian 10 (Buster)

Hi,

I think there's a bug in recent beta builds causing Microinverters to be incorrectly reported, values are being mixed up.  I updated to build 14975 on Thursday 19th.

At first I thought my database was corrupt as I had two power cuts yesterday. Here's the influx data, these are consecutive lines and correspond to the first powercut:
```
time                 Angle Azimuth CumulativeProduction Hardware    InstantProduction Microinverter    Panel             Power Sensor  Software Type
----                 ----- ------- -------------------- --------    ----------------- -------------    -----             ----- ------  -------- ----
2023-01-20T10:58:45Z 24.6  86      7077.2599            PiB+#3      111               GE2 122117035148 JA Solar JAM60S20 380   Envoy-S Domoticz IQ7A
2023-01-20T11:26:22Z 24.6  86      7128.3507            PiB+#3      271               GE2 122117035148 JA Solar JAM60S20 380   Envoy-S Domoticz IQ7A

```
Only one of my south facing panels could have been producing 271 watts, inverter 122117035148 faces east and is new.  But the problem is worse that, I've not seen my east panels get over 100 watts yet and checking the Envoy /api/v1/production/inverters today even 111 watts before the power cut is indeed wrong:
```
{
    "serialNumber": "122117035148",
    "lastReportDate": 1674297300,
    "devType": 1,
    "lastReportWatts": 33,
    "maxReportWatts": 94
  },
```
111 watts before the power cut has to have been coming from a west facing panel (also new) which have max values up to 139.  I backup every night so restored to Friday 03:00, that's at a time I was running 14975 and before the power cuts.

Still something is wrong.  Today is cloudy with very diffuse light, all my panels are producing near enough the same, I can't be sure if 122117035148 is getting the correct values so how do I know something is stilll wrong?  /api/v1/production/inverters includes two Microinverters that are decommissioned, I have 13 active and two inactive, all 15 are reported every scan by the Envoy.  Here's what the Envoy says about a decommissioned Microinverter:
```
  {
    "serialNumber": "122027007811",
    "lastReportDate": 0,
    "devType": 1,
    "lastReportWatts": 0,
    "maxReportWatts": 0
  },
```
Domoticz knows about all 15, only 13 are added.  One of my active Microinverters (122117033378) is in Domoticz reporting zero watts despite Envoy saying:
```
{
    "serialNumber": "122117033378",
    "lastReportDate": 1674297311,
    "devType": 1,
    "lastReportWatts": 35,
    "maxReportWatts": 129
  }
```
So I thought maybe my database had been corrupted by the update to 14975 so restored to before that update, Domoticz still reports zero watts for Microinverter 122117033378.  Last roll of the dice I've updated to 14980, no change.
So when did the problem get introduced?  I always update via the script so here's my backups folder:
```
-rw-r--r-- 1 root root 18227938 Dec 20 10:43 domoticz_backup_20221220_104206.tar.gz
-rw-r--r-- 1 root root 18240001 Jan  2 08:01 domoticz_backup_20230102_080045.tar.gz
-rw-r--r-- 1 root root 18285879 Jan 10 10:27 domoticz_backup_20230110_102602.tar.gz
-rw-r--r-- 1 root root 18290193 Jan 19 14:45 domoticz_backup_20230119_144451.tar.gz
-rw-r--r-- 1 root root 18005063 Jan 21 11:16 domoticz_backup_20230121_111515.tar.gz
```
Prior to the 19th 122117033378 was reporting correctly, presumably they all were 
![image](https://user-images.githubusercontent.com/58229603/213865103-33623f91-78fb-45de-8714-4121a30f504c.png)

TLDR:
- The build I installed on Jan 10th was fine, the suspect change has to be in build 14969.  
- The problem exists for sure from 14975.
- The bug is causing data for a decommissioned Microinverter (two serial numbers possible) to be logged for 122117033378
- The bug was causing data for a West facing Microinverter (seven serial numbers possible) to be logged for 122117035148
- After the powercut yesterday I had a South facing Microinverter (two serial numbers possible) to be logged for 122117035148

Cheers
SumDum

**Comment by SumDum** [comment]
How is the serialNumber read from the Enphase API matched to the Meter?  The problem could be my renaming of the devices:
![image](https://user-images.githubusercontent.com/58229603/216280051-8e3f13be-94d6-4d7f-aad2-263618d0de09.png)
As you can see I added location info into the device names

- GrX = Ground (south facing, should always have the highest production at this time of year.)
- GEx = Garage East
- GWx = Garage West

I've updated a second Pi Zero to the beta and added the microinverters. Taking for example 1222101029000 in this pi it is producing 57 watts.  That's the correct figure, it's one of only 2 south facing panels so should closely match 122101023471.  In the pi above x9000 is only producing 6 watts.
![image](https://user-images.githubusercontent.com/58229603/215995787-4a315552-7855-4fc8-8464-a4d44c1da874.png)
The 'Inv:' at the start is new, when I added the devices in the first pi above the default device names were just bare serial numbers.  Everything worked fine with my edited names until I installed 14975.  

I've tried rolling back to the backup taken on 19th Jan & the data is still not correctly assigned to devices.  When domoticz iterates through the array of json data from api/v1/production/inverters how does it match which meter to assign the data point to?

**Comment by gizmocuz** [comment]
@SumDum , Did you had a look in the EnphaseAPI.cpp source code file?
(Specially in the function getInverterDetails?)

**Comment by gizmocuz** [comment]
This is based on the received order or it might be sorted internally by jsoncpp
But you can not change the inverter name in Enphase? Or did you do this?
Anyway, this should stay the same, so if you do not touch this anymore and rename in domoticz, you should be fine

**Comment by SumDum** [comment]
> @SumDum , Did you had a look in the EnphaseAPI.cpp source code file? (Specially in the function getInverterDetails?)

Yes but it's on the limit or past the limit of my understanding what is going on.  It looks like `if (m_bFirstTimeInvertedDetails)` will run for new inverters (with this code path extant) or it would have run the first time domoticz with this code ran.  Is that correct?

**Comment by SumDum** [comment]
> This is based on the received order or it might be sorted internally by jsoncpp 
> But you can not change the inverter name in Enphase? Or did you do this? 

No, I changed the device names in domoticz. I would not have the first idea how to change the "serialNumber": Enphase reports.

> Anyway, this should stay the same, so if you do not touch this anymore and rename in domoticz, you should be fine

It was fine with my edited device names until (if my understanding in the comment above is correct) the new code ran one time.

**Comment by gizmocuz** [comment]
Not quite, the boolean 'm_bFirstTimeInvertedDetails' will make sure that the device type is set to a producing device.
I have made a test application that parses the returned inverter details, and I can tell you that we handle it in the order returned by the envoy.
So the ID's will always be the same "if this order stays the same"
And I am pretty confident that this order will be the same

I also compared the returned inverter data from 2 months ago against the returned result just now.
The order is the same. So it is not ordered by serial_number, or timestamp.

Conclusion, nothing changed and nothing is wrong.

You should have consistent data now.

Do you still see a issue?

**Comment by SumDum** [comment]
Yes.  I'll post more screenshots tomorrow, the forecast is for good weather so there should be a clear differential between what the 3 arrays produce.

What would happen if the order did not stay the same?  Is the data always assigned to the devices in the order they were first discovered?  Or is serialNumber used to pattern match against the device names each time?

**Comment by gizmocuz** [comment]
I am pretty sure that new inverters are 'added' to the list in Enphase.
I can tell you by the end of the month.
As said already, the list from Enphase is not ordered by serial or time and is 100% not random.
.... meaning there is no issue, and you should have no issue.
(I don't have an issue, and there should be more users with an issue if there was an issue)

**Comment by SumDum** [comment]
I can say with certainty they are not always added to the list.  My system history is:
18 Oct 2020 - 2 south facing panels fitted with IQ7+ inverters.
04 Mar 2021 - Inverters updated to IQ7A for higher peak production.
05 Dec 2022 - East (4) and West (7) arrays added.
All these dates are recorded in the Envoy, I'm not working from memory or notes.

All 15 inverters including the two removed in Mar 2021 are reported every scan.  If new inverters were always added the two retired inverters would still be top of the list, they are positions 3 & 10.  The 2021 new inverters would be positions 3 & 4 in the list, they are 1 & 13.  The two removed inverters are marked as deleted in the Envoy, I have no idea why it goes on reporting them.

**Comment by gizmocuz** [comment]
Okey that is a different story! You say you replaced the (faulty) inverters? Coz in this case I can indeed imagine a shift!
What I can do is use the last digits of the serials as the ID, the first seems to be fixed

**Comment by SumDum** [comment]
The removed inverters are fine, I only changed them as I could get an extra 50 watts or so out of the early panels.  IQ7A were not available over here in Dec 2020.
Is there any way to tell from my database what order the inverters were parsed from the json?  I have a daily backup so I have all logged data going back to when I installed the beta Dec 2022, I mean I can reassemble the detailed 5 day data.  If I can tell from the database I can tell you how often the json order changes and whether it was the powercuts that caused it

**Comment by gizmocuz** [comment]
This was usefull information to have mentioned at the start, that you 'replaced' inverters.
I will make a patch tomorrow for this. This will mean that everyone using this will get new results, but it is for the better in the end

**Comment by SumDum** [comment]
Thanks and apologies it took so long to extract the right info from me!  I did mention in my first post that there were 2 decommissioned inverters in the json!  Domoticz couldn't and doesn't need to know the panel was reused.  Over the life of a large system inverter failure is inevitable I guess so dead entries anywhere in the json is something that will need to be handled.

**Comment by FireWizard52** [comment]
HI, @gizmocuz 

Sorry for jumping in, but I followed this topic and commented on the forum as well.

But you said:

> What I can do is use the last digits of the serials as the ID, the first seems to be fixed.

I said there:

> How far is it fixed.
> 
> If I look to y> How far is it fixed.
> 
> If I look to @SumDum 's serials, I see that all his serial numbers are starting with 122 and the remaining 9 numbers are variable, while my serials are starting with 121 and also the remaining 9 are variable.
> My panels/inverters are from 2019 and his are a little bit younger (2020, 2021 and 2022).
> 
> We might expect that in the future the serial numbers will go up and then we will see serials starting with 123, 124 or 125.
> How are the serials of the new IQ8 inverter?
> 
> I would suggest to take the complete serial as the ID and not the last digits (how many?) of it.

**Comment by gizmocuz** [comment]
@FireWizard52 , of course this crossed my mind (most important is the last 7 digits), but I was thinking about using the default function '**SendKwhMeter**'
But to be able to use the full serial number, we have to poke directly in the database...
I have a patch waiting here, just waiting for some more generation for testing

**Comment by SumDum** [comment]
I agree with FireWizard, use the complete serial number.  That was at least you'll know there will never be a clash.

The sun is shining so here's more screenshots.  First the Pi B+ that has been running the beta since December and has values incorrectly assigned.
![image](https://user-images.githubusercontent.com/58229603/216810552-a3efc022-ec4a-4f23-9ed2-6500220a015a.png)
And the Pi Zero W that's only been running the beta a week or so.  The widgets are ordered the same as above.  Highest production from the two south facing ground mounted panels, less from the 4 East facing panels, very little from the West array of 7 and nothing at all from the two retired inverters.
![image](https://user-images.githubusercontent.com/58229603/216810596-e4ed4944-04a7-42c0-89dc-b0c2ed5ac244.png)

**Comment by gizmocuz** [comment]
Is it possible to test beta 15024? This now uses the serial number as deviceID.
You need to replace your old inverters in Domoticz with the new one (edit old inverter, choose replace device)

**Comment by SumDum** [comment]
That looks better thanks:
![image](https://user-images.githubusercontent.com/58229603/216812884-33e359fc-5279-40f8-833d-ac17ee20204d.png)
Is it safe for me to rename the new devices as I had them before?  Change 'Inv ' to 'GrE ' etc.
I have to go out now but I'll keep a close eye on it when I can & report back on Tuesday.

**Comment by FireWizard52** [comment]
Hello @gizmocuz 

I successfully installed 15024 and discovered no issues. (Of course it was already working for me)
I did not replace any device, but installed from scratch. 

Regards

**Comment by gizmocuz** [comment]
Thanks for the feedback! Great it's working

@SumDum , as mentioned, you have to replace your 'old' devices with the 'new' one.
This edit on a GRE device, choose 'replace' and select the correct 'INV' device (and after them maybe rename them)
Instead of adding them to your system (I do not see a good use for this to monitor this every day), you can place them in a group, or create a custom template page with a 'view' of how they are arranged with their values
